### ISSUE TITLE
FR: Warn when using await on PromiseLike

### ISSUE THREAD

**Issue by kwasimensah** [microsoft/TypeScript#39387]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**

PromiseLike await

**Code**

```ts
const promiseA = new Promise(resolve => {
  resolve(5);
})

async function main() {
  const a = await 5; // typescript will give warning 80007
  const b = await promiseA; // expected
  const c = await ({} as PromiseLike<number>); // The cast to PromiseLike silences 80007
}
```

**Expected behavior:**

There should be warning on the line with const c. I found this while using chai-as-promised. "await expect(promise).to.eventually.be(arg)" was throwing UnhandledPromiseRejectionWarning.  Chai.PromisedAssertion is a PromiseLike and not a Promise (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/chai-as-promised/index.d.ts#L105) and it's not safe to use with await. But the type system hid that from me.

**Actual behavior:**

No warning. 

**Playground Link:** 

https://www.typescriptlang.org/play/index.html?ssl=10&ssc=2&pln=1&pc=1#code/MYewdgzgLgBADgJxAWwJYQKYEEYF4ZgYDuMACkmpgBQIYQgA2AbhngHwwDeAUDDLfWYYqAVgCUAbm4BfMd24BDCAE8wwGADMArmqipwMZAtRgqYrrxihIsBXhgKix2CKl9r0GACN7j5-Ap0bDcrcE91fD9UWCpOaQcIMkDMABlUAGsMAB4wLWQvDAQ2SXlpIA

**Related Issues:** 

https://github.com/microsoft/TypeScript/issues/35419

**Comment by DanielRosenwasser** [comment]
Thoughts @rbuckton?

**Comment by rbuckton** [comment]
I'm not sure I understand why you would issue a warning on `await` on `PromiseLike`. That is expressly a feature of ECMAScript's `await` keyword and allows the adoption of foreign (i.e., non-native) promises. If you try to await a "thenable" (i.e., something with a callable `then` that does not conform to the definition of `PromiseLike`), you already receive an error:

https://www.typescriptlang.org/play/index.html#code/MYewdgzgLgBADgJxAWwJYQKYEEYF4ZgYDuMACkmpgBQIYQgA2AbhngHwwDeAUDDLfWYYqAVgCUAbm4BfMd26owUDAgBmAQ2CsAKgAsMYdQCMGrHnyj6wVSTPnqIATzDAYqgK4uoqcDGTrFGy5eGFBIWHU8GHUiANgRKT4w6BgjKJi4+Ap0bETQ8BTXfAzUWCpOaWiIMmzMABlUAGsMAB4wd2QjFTZbJILYABN02NKYcsqHGD0DY1NbaSA

![image](https://user-images.githubusercontent.com/3902892/86420900-e0bac000-bc8c-11ea-88ee-612a7ed13545.png).

I'm not familiar with `chai`'s `PromisedAssertion` or any of its caveats, but if it's not a valid promise it probably shouldn't be typed as one.

**Comment by kwasimensah** [comment]
1. Chai should use proper promises (in fact the extension that’s causing this seems dead) 

2. There’s nothing in the PromiseLike type that seems to actually guarantee that it works with await. In fact looking at the Javascript spec for await (https://tc39.es/ecma262/#sec-async-function-definitions) and its definition of what a Promise is (https://tc39.es/ecma262/#sec-ispromise) I don’t think custom Promise types are actually compatible with await/async and warning 800007 should actually fire for anything that’s not a native promise. Bluebird documents that’s it’s not compatible with async/await at https://github.com/petkaantonov/bluebird/issues/1434#issuecomment-320586033

**Comment by rbuckton** [comment]
The `PromiseLike` type was specifically crafted to work with async/await per the spec. What a non-native promise loses over a native promise is that there is an extra turn of the microtask queue and that async functions only return native promises. The fact that native promises take one less turn is a recent addition as well.

All that `await` cares about is that the object has a callable `then` that can accept two callback arguments, and that is specifically what `PromiseLike` in TypeScript was designed to represent.

**Comment by kwasimensah** [comment]
Sorry for the noise. Turns out there's some brittle stuff with proxying that muddles the typing here. https://github.com/domenic/chai-as-promised/issues/266#issuecomment-653518907
